### ISSUE TITLE
Tokenizing Prompts becomes extremely slow when approaching finish

### ISSUE THREAD

**Issue by SHUMKASHUN** [axolotl-ai-cloud/axolotl#1416]
### Please check that this issue hasn't been reported before.

- [X] I searched previous [Bug Reports](https://github.com/OpenAccess-AI-Collective/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

The Tkenizing Prompts has a relatively consistent speed.

### Current behaviour

At the initial stage, the tokenizing is quite fast with ~ 5000 examples /s , but when times goes, it becomes slower and slower, until reach 8 example/s.
<img width="948" alt="Screenshot 2024-03-17 at 3 18 07 AM" src="https://github.com/OpenAccess-AI-Collective/axolotl/assets/31779666/7aced3bb-44c8-4f2f-8cf7-491139360680">

### Steps to reproduce

1. Prepare the completion type json dataset
2. using the exact same environment as in README to run the train script on Mistral model.

### Config yaml

```yaml
base_model: mistralai/Mistral-7B-v0.1
model_type: MistralForCausalLM
tokenizer_type: LlamaTokenizer

load_in_8bit: false
load_in_4bit: false
strict: false

datasets:
  - path: ****/pretrain/MathPretrain/preprocess/output_for_axo.jsonl
    type: completion
dataset_prepared_path: ****/axolotl/examples/mistral/last_run_prepared/
val_set_size: 0
output_dir: ./out

sequence_len: 4096
sample_packing: true
pad_to_sequence_len: false
eval_sample_packing: false

wandb_project: axo_conti_pretrain_mathpile
wandb_entity:
wandb_watch:
wandb_name: mistral7B_test
wandb_log_model:

gradient_accumulation_steps: 1
micro_batch_size: 1
num_epochs: 1
optimizer: adamw_hf
lr_scheduler: cosine
learning_rate: 0.000005

train_on_inputs: false
group_by_length: true
bf16: true
fp16:
tf32: false

chat_template:
sample_packing_eff_est: 1.0

gradient_checkpointing: true
early_stopping_patience:
resume_from_checkpoint:
local_rank:
logging_steps: 1
xformers_attention:
flash_attention: true

warmup_steps: 10
evals_per_epoch: 0
eval_table_size:
eval_max_new_tokens: 128
save_steps: 20
debug:
deepspeed: ****/axolotl/deepspeed_configs/zero3_bf16.json
weight_decay: 0.0
fsdp:
fsdp_config:
special_tokens:
  bos_token: "<s>"
  eos_token: "</s>"
  unk_token: "<unk>"
```

### Which Operating Systems are you using?

- [X] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.10

### axolotl branch-commit

main

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this bug has not been reported yet.
- [X] I am using the latest version of axolotl.
- [X] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by tmm1** [comment]
Can you check system stats such as memory and swap usage?

**Comment by SHUMKASHUN** [comment]
Thanks for the reply. Yes, I have monitored the memory usage, I am the only person who is using the machine, and the tokenizing process did not consume much memory, where I have huge free memory ~ 1T.

**Comment by NanoCode012** [comment]
This is stale but was an issue till recently.

We've implemented a few fixes for this:
- reducing num tokenization for chat_templates by half
- used proper vcpu count based on systems and removed limiter
